### PR TITLE
Use VW pool as the only way for getting clients

### PR DESCRIPTION
**Pull request by zeromem**
I hereby agree to the terms of the CLA available at: [placeholder]

Changelog category (leave one):
- Improvement

Detailed description / Documentation draft:
- Use VirtualWarehousePool as the only way for getting workers.
- Fix divide zero error for ResourceTracker

If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ByConity Documentation](https://github.com/ByConity/ByConity/blob/master/CONTRIBUTING.md) guide first.

### CODE REVIEW

File: src/TableFunctions/TableFunctionCnch.cpp [hunk 1]
```diff
@@ -36,22 +37,11 @@ namespace ErrorCodes
...
-    catch(const Exception & e)
-    {
-        throw Exception("Get worker clients for vw " + vw_name + " failed. Got exception "
-             + e.displayText(), e.code());
-    }
+    auto vw_handle = context.getVirtualWarehousePool().get(vw_name);
+    auto value = context.getSettingsRef().vw_schedule_algo.value;
+    auto algo = ResourceManagement::toVWScheduleAlgo(&value[0]);
+    auto wg_handle = vw_handle->pickWorkerGroup(algo);
+    auto & worker_clients = wg_handle->getWorkerClients();
```
Comment (yuanquan404) [inline src/TableFunctions/TableFunctionCnch.cpp:44]: 这里不能使用引用，getAllWorkers返回值是临时对象

File: src/Interpreters/VirtualWarehouseHandle.cpp [hunk 2]
```diff
@@ -339,6 +339,35 @@ WorkerGroupHandle VirtualWarehouseHandleImpl::pickLocally(const VWScheduleAlgo &
...
+    return wg_handle->getWorkerClient();
+}
+
+std::vector<CnchWorkerClientPtr> VirtualWarehouseHandleImpl::getAllWorkers()
+{
+    std::vector<CnchWorkerClientPtr> res;
+    std::lock_guard lock(state_mutex);
+    for (auto const & [_, wg_handle] : worker_groups)
+    {
+        auto & workers = wg_handle->getWorkerClients();
```
Comment (yuanquan404) [inline src/Interpreters/VirtualWarehouseHandle.cpp:365]: remove reference

File: src/Interpreters/WorkerGroupHandle.cpp [hunk 3]
```diff
@@ -140,6 +140,26 @@ WorkerGroupHandleImpl::WorkerGroupHandleImpl(const WorkerGroupHandleImpl & from,
     }
 }
 
+CnchWorkerClientPtr WorkerGroupHandleImpl::getWorkerClientByHash(const String & key) const
+{
+    UInt64 index = std::hash<String>{}(key);
+    if (worker_clients.empty() || index >= worker_clients.size())
+        throw Exception("No available worker for " + id, ErrorCodes::RESOURCE_MANAGER_NO_AVAILABLE_WORKER);
```
Comment (yuanquan404) [inline src/Interpreters/WorkerGroupHandle.cpp:147]: index是hash值，超过worker_clients.size()应该很正常？

File: src/TableFunctions/TableFunctionCnch.cpp (hunk 1 above)
Comment (zeromem) [reply src/TableFunctions/TableFunctionCnch.cpp:44]: @yuanquan404 我看后面是为了提取worker client里面的address来构造Cluster对象。这里只要wg_handle没有释放，应该就没问题？

File: src/Storages/Kafka/CnchKafkaConsumerScheduler.cpp [hunk 4]
```diff
@@ -131,7 +126,7 @@ namespace DB
 
             clients_status_map.erase(node.client_ptr);
         }
-        throw Exception("No available service for " + worker_pool->getServiceName(), ErrorCodes::LOGICAL_ERROR);
+        throw Exception("No available workers for scheduling Kafka consumer, vw: " + vw_name, ErrorCodes::LOGICAL_ERROR);
```
Comment (yuanquan404) [inline src/Storages/Kafka/CnchKafkaConsumerScheduler.cpp:129]: ErrorCodes 改下？

File: src/TableFunctions/TableFunctionCnch.cpp (hunk 1 above)
Comment (yuanquan404) [reply src/TableFunctions/TableFunctionCnch.cpp:44]: 不，返回值的有效期是调用的那一行

File: src/Interpreters/WorkerGroupHandle.cpp (hunk 3 above)
Comment (zeromem) [reply src/Interpreters/WorkerGroupHandle.cpp:147]: 一开始这里方法本来是想做成 getWorkerClientByIndex()。 我改下

**Comment by canhld94** [comment]
@zeromem has this one ready for review?

**Comment by zeromem** [comment]
> @zeromem has this one ready for review?

yeah please help to review. And please help to take a look at the CI issue, there is always a timeout testcase.

**Comment by canhld94** [comment]
> > @zeromem has this one ready for review?
> 
> yeah please help to review. And please help to take a look at the CI issue, there is always a timeout testcase.

CI issue will be fix soon. Btw, the bug makes RM crash, does the fix included in this PR?

File: src/TableFunctions/TableFunctionCnch.cpp [hunk 5]
```diff
@@ -36,27 +37,16 @@ namespace ErrorCodes
...
-    }
-    catch(const Exception & e)
-    {
-        throw Exception("Get worker clients for vw " + vw_name + " failed. Got exception "
-             + e.displayText(), e.code());
-    }
+    auto vw_handle = context.getVirtualWarehousePool().get(vw_name);
+    auto value = context.getSettingsRef().vw_schedule_algo.value;
+    auto algo = ResourceManagement::toVWScheduleAlgo(&value[0]);
+    auto wg_handle = vw_handle->pickWorkerGroup(algo);
```
Comment (canhld94) [inline src/TableFunctions/TableFunctionCnch.cpp:43]: Here we only get all workers in a vw, not just a wg. May be we add a new method `getAllWorkers` to virtualware pool?

Comment (zeromem) [reply src/TableFunctions/TableFunctionCnch.cpp:43]: Actually I added a method 'getAllWorkers' in VirtualWarehouseHandle and we can change the implementation here. But in a vw with multi wg, each query is always processed by one wg.

**Review by canhld94 — APPROVED** [review]

**Comment by canhld94** [comment]
Fail test doesn't relates

**Pull request merged**